File: sdks/python/aegis_webhooks/verify.py
"""HMAC-SHA256 signature verification for Aegis webhook deliveries.

Mirrors the signing logic in backend/src/notifications/webhook_signing.py so
receivers can validate authenticity and replay protection without rolling their
own crypto.
"""
from __future__ import annotations

import hashlib
import hmac
import json
import time
from typing import Mapping


class AegisWebhookError(Exception):
    """Base class for all Aegis webhook verification errors."""


class InvalidTimestampError(AegisWebhookError):
    """Timestamp is missing, unparseable, or outside the tolerance window."""


class InvalidSignatureError(AegisWebhookError):
    """No candidate signature matched any of the provided secrets."""


def _canonical_json(payload: Mapping | bytes | str) -> str:
    """Return the canonical JSON string used in the signed string.

    WHY: Phase 44 signs `{ts}.{canonical_json}` where canonical means
    sort_keys=True and no whitespace — matches json.dumps defaults in
    webhook_signing.py::sign_payload exactly.
    """
    if isinstance(payload, (bytes, str)):
        # Re-parse so we sort keys regardless of original serialisation order
        obj = json.loads(payload)
    else:
        obj = payload
    return json.dumps(obj, sort_keys=True, separators=(",", ":"))


def _case_insensitive(headers: Mapping[str, str]) -> dict[str, str]:
    """Return a lowercase-keyed copy of headers.

    WHY: RFC 7230 §3.2 declares HTTP header names case-insensitive; frameworks
    differ (Flask lowercases, Django keeps original case, etc.).
    """
    return {k.lower(): v for k, v in headers.items()}


def verify_signature(
    payload: Mapping | bytes | str,
    secret: str | list[str],
    headers: Mapping[str, str],
    *,
    tolerance_seconds: int = 300,
    current_time: int | None = None,
) -> None:
    """Verify an Aegis webhook signature.

    Raises ``InvalidTimestampError`` or ``InvalidSignatureError`` on failure.
    Returns ``None`` on success (never returns True/False — callers should
    treat any return as success and only catch exceptions on failure).

    Parameters
    ----------
    payload:
        The raw request body as ``bytes`` / ``str``, or the already-parsed
        JSON object as a ``dict``.  If bytes/str the function re-parses it
        to produce canonical JSON.
    secret:
        A single signing secret string, or a list of secrets (pass
        ``[old, new]`` during a rotation window — verification passes if
        any secret matches).
    headers:
        The HTTP request headers.  Lookup is case-insensitive so both
        ``X-Aegis-Signature`` and ``x-aegis-signature`` are accepted.
    tolerance_seconds:
        Maximum age of a delivery in seconds.  Deliveries older or
        future-dated beyond this window are rejected to prevent replay
        attacks.  Default: 300 (5 minutes), matching the server default.
    current_time:
        Override the current Unix timestamp used for the tolerance check.
        Intended for unit testing only.
    """
    ci = _case_insensitive(headers)

    # ── Extract required headers ────────────────────────────────────────────
    timestamp_str = ci.get("x-aegis-timestamp")
    signature_header = ci.get("x-aegis-signature")

    if timestamp_str is None or signature_header is None:
        raise AegisWebhookError(
            "Missing required headers: X-Aegis-Timestamp and X-Aegis-Signature must be present"
        )

    # ── Validate timestamp ──────────────────────────────────────────────────
    try:
        ts = int(timestamp_str)
    except (ValueError, TypeError) as exc:
        raise InvalidTimestampError(
            f"X-Aegis-Timestamp is not a valid integer: {timestamp_str!r}"
        ) from exc

    now = current_time if current_time is not None else int(time.time())
    age = now - ts
    if abs(age) > tolerance_seconds:
        raise InvalidTimestampError(
            f"Timestamp is outside the tolerance window "
            f"(age={age}s, tolerance={tolerance_seconds}s)"
        )

    # ── Build signed string ─────────────────────────────────────────────────
    canonical = _canonical_json(payload)
    signed = f"{timestamp_str}.{canonical}".encode()

    # ── Collect candidate (v1=<hex>) values from the header ─────────────────
    # WHY: Rotation sends multiple comma-separated signatures; we accept any.
    candidates = [s.strip() for s in signature_header.split(",") if s.strip()]

    # ── Verify against each secret × each candidate signature ───────────────
    secrets_list = [secret] if isinstance(secret, str) else secret
    for raw_secret in secrets_list:
        expected_hex = hmac.new(
            raw_secret.encode(), signed, hashlib.sha256
        ).hexdigest()
        expected = f"v1={expected_hex}"
        for candidate in candidates:
            # WHY: hmac.compare_digest prevents timing-based secret leakage.
            if hmac.compare_digest(expected, candidate):
                return  # Success — at least one pair matches

    raise InvalidSignatureError(
        "No candidate signature matched any of the provided secrets"
    )

File: sdks/node/src/verify.ts
/**
 * HMAC-SHA256 signature verification for Aegis webhook deliveries.
 *
 * Mirrors the signing logic in backend/src/notifications/webhook_signing.py so
 * receivers can validate authenticity and replay protection without rolling
 * their own crypto.
 */
import { createHmac, timingSafeEqual } from "crypto";

export class AegisWebhookError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "AegisWebhookError";
  }
}

export class InvalidTimestampError extends AegisWebhookError {
  constructor(message: string) {
    super(message);
    this.name = "InvalidTimestampError";
  }
}

export class InvalidSignatureError extends AegisWebhookError {
  constructor(message: string) {
    super(message);
    this.name = "InvalidSignatureError";
  }
}

export interface VerifyOptions {
  /** Maximum age of a delivery in seconds. Default: 300 (5 minutes). */
  toleranceSeconds?: number;
  /**
   * Override the current Unix timestamp used for the tolerance check.
   * Intended for unit testing only — allows freezing time without mocking.
   */
  currentTime?: number;
}

/**
 * Return a lowercase-keyed lookup from arbitrary header maps.
 *
 * WHY: RFC 7230 §3.2 declares HTTP header names case-insensitive; different
 * frameworks (Express, Fastify, Hono, …) may preserve original casing.
 */
function normaliseHeaders(headers: Record<string, string | string[] | undefined>): Map<string, string> {
  const map = new Map<string, string>();
  for (const [key, value] of Object.entries(headers)) {
    if (value === undefined) continue;
    // Take the first value if the framework provides an array (e.g. multi-value)
    map.set(key.toLowerCase(), Array.isArray(value) ? value[0] : value);
  }
  return map;
}

/**
 * Return canonical JSON matching Phase 44's sign_payload serialisation.
 *
 * WHY: Keys must be sorted alphabetically with no whitespace so the signed
 * string is byte-for-byte identical to what the server produced.
 */
function canonicalJson(payload: object): string {
  // Deep-clone via JSON round-trip so we can sort keys at every nesting level
  const clone = JSON.parse(JSON.stringify(payload)) as object;
  return JSON.stringify(clone, Object.keys(clone).sort() as never);
}

/**
 * Recursively sort the keys of a JSON object so nested objects are also sorted.
 * Used to produce canonical JSON that matches Python's json.dumps(sort_keys=True).
 */
function sortObjectKeys(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(sortObjectKeys);
  if (value !== null && typeof value === "object") {
    const obj = value as Record<string, unknown>;
    return Object.keys(obj)
      .sort()
      .reduce<Record<string, unknown>>((acc, k) => {
        acc[k] = sortObjectKeys(obj[k]);
        return acc;
      }, {});
  }
  return value;
}

/**
 * Return canonical JSON with all object keys sorted recursively.
 *
 * WHY: Phase 44 uses Python's sort_keys=True which sorts at every nesting
 * level; a single top-level sort is insufficient for nested objects.
 */
function canonicalJsonDeep(payload: object | string | Buffer): string {
  let obj: unknown;
  if (Buffer.isBuffer(payload)) {
    obj = JSON.parse(payload.toString("utf8"));
  } else if (typeof payload === "string") {
    obj = JSON.parse(payload);
  } else {
    obj = payload;
  }
  return JSON.stringify(sortObjectKeys(obj));
}

/**
 * Compute v1=<hex> for a given secret and signed string.
 *
 * WHY: Extracted to keep the comparison loop readable and avoid repeating the
 * HMAC construction.
 */
function computeSignature(secret: string, signedString: string): string {
  const hex = createHmac("sha256", secret)
    .update(signedString, "utf8")
    .digest("hex");
  return `v1=${hex}`;
}

/**
 * Verify an Aegis webhook signature.
 *
 * Throws `InvalidTimestampError` or `InvalidSignatureError` on failure.
 * Returns `void` on success.
 *
 * @param payload  Raw request body as `Buffer` / `string`, or already-parsed
 *                 JSON object.  Bytes/strings are re-parsed to produce canonical JSON.
 * @param secret   A single signing secret string, or an array of secrets
 *                 (pass `[oldSecret, newSecret]` during a rotation window).
 * @param headers  The HTTP request headers.  Lookup is case-insensitive.
 * @param options  Optional tolerance window and time override for testing.
 */
export function verifySignature(
  payload: object | string | Buffer,
  secret: string | string[],
  headers: Record<string, string | string[] | undefined>,
  options?: VerifyOptions,
): void {
  const { toleranceSeconds = 300, currentTime } = options ?? {};
  const normalised = normaliseHeaders(headers);

  // ── Extract required headers ────────────────────────────────────────────
  const timestampStr = normalised.get("x-aegis-timestamp");
  const signatureHeader = normalised.get("x-aegis-signature");

  if (timestampStr === undefined || signatureHeader === undefined) {
    throw new AegisWebhookError(
      "Missing required headers: X-Aegis-Timestamp and X-Aegis-Signature must be present",
    );
  }

  // ── Validate timestamp ──────────────────────────────────────────────────
  const ts = Number(timestampStr);
  if (!Number.isInteger(ts) || String(ts) !== timestampStr) {
    throw new InvalidTimestampError(
      `X-Aegis-Timestamp is not a valid integer: ${JSON.stringify(timestampStr)}`,
    );
  }

  const now = currentTime ?? Math.floor(Date.now() / 1000);
  const age = now - ts;
  if (Math.abs(age) > toleranceSeconds) {
    throw new InvalidTimestampError(
      `Timestamp is outside the tolerance window (age=${age}s, tolerance=${toleranceSeconds}s)`,
    );
  }

  // ── Build signed string ─────────────────────────────────────────────────
  const canonical = canonicalJsonDeep(payload);
  const signedString = `${timestampStr}.${canonical}`;

  // ── Collect candidate v1=<hex> values from the header ───────────────────
  // WHY: Rotation sends multiple comma-separated signatures; we accept any match.
  const candidates = signatureHeader
    .split(",")
    .map((s) => s.trim())
    .filter(Boolean);

  // ── Verify against each secret × each candidate signature ───────────────
  const secrets = Array.isArray(secret) ? secret : [secret];
  for (const rawSecret of secrets) {
    const expected = computeSignature(rawSecret, signedString);
    const expectedBuf = Buffer.from(expected, "utf8");

    for (const candidate of candidates) {
      const candidateBuf = Buffer.from(candidate, "utf8");
      // WHY: timingSafeEqual prevents timing-based secret leakage — buffers
      // must be the same length, so length mismatch is checked first.
      if (
        expectedBuf.length === candidateBuf.length &&
        timingSafeEqual(expectedBuf, candidateBuf)
      ) {
        return; // Success
      }
    }
  }

  throw new InvalidSignatureError(
    "No candidate signature matched any of the provided secrets",
  );
}
